hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/processor/TwoPhaseSnapshotCommitUtility.java

```diff
@@ -0,0 +1,427 @@
+    }
+
+    protected Consumer<TXN_ID> recoverAndCommitFn() {
+        return recoverAndCommitFn;
+    }
+
+    protected ConsumerEx<Integer> recoverAndAbortFn() {
+        return recoverAndAbortFn;
+    }
+
+    /**
+     * Returns the active transaction that can be used to store an item or
+     * query the source. It's null in case when a transaction is not available
+     * now. In that case the processor should back off and retry later.
+     */
+    @Nullable
+    public abstract RES activeTransaction();
+
+    /**
+     * For sinks and inner vertices, call from {@link Processor#complete()}.
+     * For batch sources call after the source emitted everything. Never call
+     * it for streaming sources.
+     */
+    public abstract void afterCompleted();
+
+    /**
+     * Delegate handling of {@link Processor#saveToSnapshot()} to this method.
+     *
+     * @return a value to return from {@code saveToSnapshot()}
+     */
+    public abstract boolean saveToSnapshot();
+
+    /**
+     * Delegate handling of {@link Processor#onSnapshotCompleted(boolean)} to
+     * this method.
+     *
+     * @param commitTransactions value passed to {@code onSnapshotCompleted}
+     * @return value to return from {@code onSnapshotCompleted}
+     */
+    public abstract boolean onSnapshotCompleted(boolean commitTransactions);
+
+    /**
+     * Delegate handling of {@link Processor#restoreFromSnapshot(Inbox)} to
+     * this method. If you save custom items to snapshot besides those saved by
+     * {@link #saveToSnapshot()} of this utility, use {@link
+     * #restoreFromSnapshot(Object, Object)} to pass only entries not handled
+     * by your processor.
+     *
+     * @param inbox the inbox passed to {@code Processor.restoreFromSnapshot()}
+     */
+    @SuppressWarnings("unchecked")
+    public void restoreFromSnapshot(@Nonnull Inbox inbox) {
+        for (Object item; (item = inbox.poll()) != null; ) {
+            Entry<BroadcastKey<TXN_ID>, Boolean> castedItem = (Entry<BroadcastKey<TXN_ID>, Boolean>) item;
+            restoreFromSnapshot(castedItem.getKey(), castedItem.getValue());
+        }
+    }
+
+    /**
+     * Delegate handling of {@link
+     * AbstractProcessor#restoreFromSnapshot(Object, Object)} to this method.
+     *
+     * <p></p>See also {@link #restoreFromSnapshot(Inbox)}.
+     *
+     * @param key a key from the snapshot
+     * @param value a value from the snapshot
+     */
+    public abstract void restoreFromSnapshot(@Nonnull Object key, @Nonnull Object value);
+
+    /**
+     * Call from {@link Processor#close()}.
+     *
+     * <p>The implementation must not commit or rollback any pending transactions
+     * - the job might have failed between after snapshot phase 1 and 2. The
+     * pending transactions might be recovered after the restart.
+     */
+    public abstract void close() throws Exception;
+
+    protected boolean usesTransactionLifecycle() {
+        return externalGuarantee == EXACTLY_ONCE ||
```

Comment:
I see these two variables only used here, but this method is not used. Are these needed?

Reply:
The whole utility is unused here. Also it's abstract. It is used in `2pc-file` and `2pc-kafka` branches, in both of them.